### PR TITLE
tls_codec: Fix variable length encoding

### PR DESCRIPTION
**Pull request by pcapriotti**
- Lengths over 2 bytes were serialised with an incorrect prefix
- Fixed off-by-one errors on encoding length boundaries
- Length prefix should be the logarithm of the number of bytes to read to extract the length

### CODE REVIEW

File: tls_codec/src/quic_vec.rs [hunk 1]
```diff
@@ -76,9 +85,9 @@ fn length_encoding_bytes(length: u64) -> Result<usize, Error> {
 
     Ok(if length < 0x40 {
         1
-    } else if length < 0x3fff {
+    } else if length < 0x4000 {
```
Comment (franziskuskiefer) [inline tls_codec/src/quic_vec.rs:88]: Nice catch. I'd prefer `<=` for these and keep the values.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Good catch. This hasn't been tested very well.
Can you add test cases for all the edge cases around the boundaries?

The deserialize is actually missing a check that the minimum encoding was used if `feature = "mls"`. Can you add that as well while you're at it?

**Comment by pcapriotti** [comment]
I added boundary tests and a check that the minimum encoding is used.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Thanks, one more thing, then we can get this in.

File: tls_codec/src/quic_vec.rs [hunk 2]
```diff
@@ -49,19 +49,34 @@ fn read_variable_length<R: std::io::Read>(bytes: &mut R) -> Result<(usize, usize
...
+
+    for _ in 1..len_len {
         let mut next = [0u8; 1];
         bytes.read_exact(&mut next)?;
         length = (length << 8) + usize::from(next[0]);
     }
 
+    // ensure that len_len is minimal for the given length
+    let min_len_len = length_encoding_bytes(length as u64)?;
+    if min_len_len != len_len {
```
[the remark this answers is not in the document]
Comment (pcapriotti) [reply tls_codec/src/quic_vec.rs]: Yep, sorry, I forgot to add the feature check. Fixed now.

**Review by franziskuskiefer — APPROVED** [review]
Thanks!
